### PR TITLE
[Impeller] hash even less stuff per frame.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Follow up to https://github.com/flutter/engine/pull/55060

Currently we have multiple stages of hashing while font rendering, which is relatively expensive for the actualy required workload. First, we hash the contents of all text frames to compute the unique set of glyphs per frame. Then we diff this hash set against the hashmap of glyphs within the atlas. Finally we hash and lookup the final rendered bounds for each glyph.

We can simplify this to 2. hash lookups for glyphs not yet in the atlas and 1. hash lookup for glyphs that are in the atlas. This is done by combing the step where we uniquely compute glyphs per frame with the diff against the current atlas. When this lookup is performed, we also store the glyph position (if found) in the text_frame itself - which allows text contents to skip the last hash, as long as the glyph has already been rendered.

### Before

flutter_03

### After

flutter_04

Using this handy dandy test app:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {

  Widget build(context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Platform View'),
        ),
        body: SafeArea(child: Stack(children: [
          SizedBox(
            width: 380,
            height: 380,
            child: LinearProgressIndicator(),
          ),
          Stack(
            children: List<Widget>.generate(1000, (index) {
              // The problem already happens with a small amount of widgets.
              // Using an excessive amount of widgets is just to make the problem more evident.
              return Text("Lots of Texts represent a Widget with complex components.");
            }),
          ),

          Align(
            alignment: Alignment.bottomCenter,
            child:
            TextButton(
              child: Text("Button"),
              onPressed: () {
                print("Tap ${DateTime.now()}");
              },
            ),
          ),
        ],
        ),
        ),
      ),
    );
  }
}
```

### CODE REVIEW

File: impeller/typographer/text_frame.h
Comment (jonahwilliams) [inline impeller/typographer/text_frame.h:19]: If this behavior is a problem, we could instead use a different object as a cache, but I think that would be substantially more refactoring for little to no gain, as this condition is always true for us.

File: impeller/typographer/backends/skia/typographer_context_skia.cc
Comment (jonahwilliams) [inline impeller/typographer/backends/skia/typographer_context_skia.cc:414]: Leaving this as the next optimizations. With a little plumbing when we recreate the atlas we can actually avoid any hashing on frames with no new text.

File: impeller/typographer/text_frame.h
Comment (jtmcdole) [inline impeller/typographer/text_frame.h:110]: super nit for clarity: "per frame data" -> "per text frame data"

**Review by jtmcdole — APPROVED** [review]
lgtm from me; but you asked for Chinmay

**Comment by jonahwilliams** [comment]
FYI @chinmaygarde this is still waiting on a review from you.

File: impeller/typographer/text_frame.h
Comment (chinmaygarde) [inline impeller/typographer/text_frame.h:114]: `= std::nullopt` is redundant.

File: impeller/typographer/text_frame.h
Comment (chinmaygarde) [inline impeller/typographer/text_frame.h:112]: Zero out the scalar.

File: impeller/typographer/text_frame.h
Comment (chinmaygarde) [reply impeller/typographer/text_frame.h:19]: Sounds good and thanks for the clarification.

File: impeller/typographer/text_frame.cc
Comment (chinmaygarde) [inline impeller/typographer/text_frame.cc:124]: If you use `.at(index)` you get bounds checking in debug modes for free.

Comment (chinmaygarde) [reply impeller/typographer/text_frame.cc:124]: GetFrameBounds can be `const`.

File: impeller/typographer/glyph_atlas.h
Comment (chinmaygarde) [inline impeller/typographer/glyph_atlas.h:29]: Zero out the scalar.

File: impeller/typographer/glyph_atlas.h
Comment (chinmaygarde) [inline impeller/typographer/glyph_atlas.h:246]: Did you mean to comment this out? Couldn't find where you don't want to copy but do want to make it copy assigned. At least the rule of 5 thing doesn't track.

File: impeller/typographer/glyph_atlas.h
Comment (chinmaygarde) [reply impeller/typographer/glyph_atlas.h:29]: Name this `is_placeholder` instead? While reading this later, I thought placeholder returned a placeholder itself.

File: impeller/typographer/backends/stb/typographer_context_stb.cc
Comment (chinmaygarde) [inline impeller/typographer/backends/stb/typographer_context_stb.cc:472]: DCHECK that the sizes of the two out vectors are the same?

Comment (chinmaygarde) [reply impeller/typographer/backends/stb/typographer_context_stb.cc:472]: I'm slightly partial to returning a pair of vectors instead of out variables because it makes it clear that existing values in the vectors will be not be disregarded. I didn't see a place where you didn't have to clear the vectors before calling CollectNewGlyphs. Just one less thing you don't have to remember to do. Then you can use a structured binding to do something like:

```c++
auto [new_glyphs, new_sizes] = CollectNewGlyphs(last_atlas, text_frames);
```

File: impeller/typographer/backends/skia/typographer_context_skia.cc
Comment (chinmaygarde) [reply impeller/typographer/backends/skia/typographer_context_skia.cc:414]: sgtm

**Review by chinmaygarde — APPROVED** [review]
> this is still waiting on a review from you

Sorry about the delay. All nits.

**Comment by jonahwilliams** [comment]
Thank you! :)

File: impeller/typographer/glyph_atlas.h
Comment (jonahwilliams) [reply impeller/typographer/glyph_atlas.h:246]: I forgot about this one. I wasn't sure if we should try to design around this. Basically with this I can't do:

```
  font_atlas_map_[scaled_font] = FontGlyphAtlas();
  return &font_atlas_map_[scaled_font];
```

and I'm not sure what the alternative is.

File: impeller/typographer/backends/stb/typographer_context_stb.cc
Comment (jonahwilliams) [reply impeller/typographer/backends/stb/typographer_context_stb.cc:472]: Done. I think I one point we were incrementally adding to these but that is long since gone

File: impeller/typographer/glyph_atlas.h
Comment (jonahwilliams) [reply impeller/typographer/glyph_atlas.h]: Done

File: impeller/typographer/text_frame.cc
Comment (jonahwilliams) [reply impeller/typographer/text_frame.cc]: Done

File: impeller/typographer/text_frame.h
Comment (jonahwilliams) [reply impeller/typographer/text_frame.h]: Done

File: impeller/typographer/text_frame.h
Comment (jonahwilliams) [reply impeller/typographer/text_frame.h]: Done

**Comment by jonahwilliams** [comment]
reason for revert: framework golden failures.

File: impeller/typographer/glyph_atlas.h
Comment (chinmaygarde) [reply impeller/typographer/glyph_atlas.h]: I think the alternative is [`insert`](https://en.cppreference.com/w/cpp/container/unordered_map/insert). You provide a pair and it returns an iterator.

**Pull request merged**
